src/core/qgssnappingutils.cpp

```diff
@@ -330,6 +330,57 @@ QgsPointLocator::Match QgsSnappingUtils::snapToMap( const QgsPointXY &pointMap,
 }
 
 
+QgsFeatureIds QgsSnappingUtils::isMatchAVisibleLayer( QgsVectorLayer *layer )
+{
+
+  QgsFeatureIds fids;
+
+    if ( layer )
+    {
+        if (!mEnableSnappingForInvisibleFeature) {
+      bool filter = false;
+      QgsRenderContext context( QgsRenderContext::fromMapSettings( mMapSettings ) );
+      context.expressionContext() << QgsExpressionContextUtils::layerScope( layer );
+      std::unique_ptr< QgsFeatureRenderer > renderer( layer->renderer() ? layer->renderer()->clone() : nullptr );
+      if ( renderer )
+      {
+        // setup scale for scale dependent visibility (rule based)
+        renderer->startRender( context, layer->fields() );
+        filter = renderer->capabilities() & QgsFeatureRenderer::Filter;
+      }
+
+
+      QgsFeatureIterator fi = layer->getFeatures();
```

Comment:
Use an optimised request here:

- use the renderer filter (as explored earlier) to partially limit the requested features
- set the subset of attributes to match only those required by the renderer
- maybe limit to mapSettings extent? (not sure)

Reply:
I don't think we should limit to map extent. 
ping @wonder-sk

Reply:
@3nids @wonder-sk @nyalldawson 
Thank you for your reviews.
With my last commit, I have good performance for normal case (at least my use case : cadastre, road and networks, land survey, etc.). It hangs only on very big layer (more than 100k) when is filtered.

I wonder, if mapExtent should be the solution for locators created from QgsMapCanvasSnappingUtils?

@wonder-sk 
>How about updating QgsPointLocator::init() method to have optional check of feature visibility >based on a passed feature renderer?

Can you explain more please, I'm not sure if I understand? Is mEnableSnappingForInvisibleFeature flag not do this?

Reply:
So right now in the first phase you build a set of feature IDs that will be used, then in the second phase you pass it to QgsPointLocator which uses the feature IDs in the rebuildIndex() method. What I am suggesting is to merge these two phases into one: rebuildIndex() would have access to a renderer (if enableSnappingForInvisibleFeature=false) and check its willRenderFeature() and add only visible features to the index. In this way we would avoid creation of the temporary set of feature IDs and we would not need to iterate over the layers twice.